xds/internal/balancer/cdsbalancer/cdsbalancer_test.go

```diff
@@ -398,36 +407,127 @@ func (s) TestHandleClusterUpdate(t *testing.T) {
 	}
 }
 
-// TestResolverError verifies that an existing watch is cancelled when a
-// resolver error is received by the cdsBalancer, and also that the same error
-// is propagated to the edsBalancer.
+// TestHandleClusterUpdateError covers the cases that an error is returned from
+// the watcher.
+//
+// Includes error with and without a child eds balancer, and whether error is a
+// resource-not-found error.
+func (s) TestHandleClusterUpdateError(t *testing.T) {
+	xdsC, cdsB, edsB, tcc, cancel := setupWithWatch(t)
+	defer func() {
+		cancel()
+		cdsB.Close()
+	}()
+
+	// An error before eds balancer is built. Should result in an error picker.
+	// And this is not a resource not found error, watch shouldn't be canceled.
+	err1 := errors.New("cdsBalancer resolver error 1")
+	xdsC.InvokeWatchClusterCallback(xdsclient.ClusterUpdate{}, err1)
+	if err := xdsC.WaitForCancelClusterWatch(); err == nil {
+		t.Fatal("watch was canceled, want not canceled (timeout error)")
+	}
+	if err := edsB.waitForResolverError(err1); err == nil {
+		t.Fatal("eds balancer shouldn't get error (shouldn't be built yet)")
+	}
+	if state, err := tcc.newPickerCh.Receive(); err != nil {
+		t.Fatalf("failed to get picker, expect an error picker")
+	} else {
```

Comment:
We don't need this else. The code in the else block can be unconditional.

Reply:
We need else to capture the `state` from `if`.
Or, I move `state, err := tcc.newPickerCh.Receive()` out from `if`.

Reply:
Makes sense. I missed that.